docs/2020-06-22-priority-rewrites.md

```diff
@@ -0,0 +1,229 @@
+
+Let us notice that, given a rule `φ(X) ⇒ ψ(X)`, we could describe
+all configurations to which it applies with `∃ X . phi(X)`. Therefore,
+if we have only two rules (`φᵢ` includes the requires predicate):
+```
+φ₁(X) ⇒ ψ₁(X) [priority(p₁)]
+φ₂(X) ⇒ ψ₂(X) [priority(p₂)]
+```
+and, if we also know that `p₁ < p₂`, we could replace them with
+```
+φ₁(X) ⇒ ψ₁(X)
+(φ₂(X) ∧ ¬ ∃ X₁ . φ₁(X₁)) ⇒ ψ₂(X)
+```
+
+Let us see that this is, indeed, the case. We would like to show that the
+result of applying the second rule is the same.
+
+The result of applying the second non-priority rule to `C(Y)` is
+```
+∃ X . ⌈ C(Y) ∧ (φ₂(X) ∧ ¬ ∃ X₁ . φ₁(X₁)) ⌉ ∧ ψ₂(X)
+```
+
+Applying the second rule when the first one doesn't work makes most sense for
+function-like patterns. For non-function-like patterns understanding what it
+means to apply the rules separately is slightly trickier, but it can be
+reduced to splitting the pattern into function like ones (if possible) and
+applying the rules on the functional components.
+
+If C(Y) is functional, then the first rule does not apply iff
+`¬ ∃ X₁ ⌈ C(Y) ∧ φ₁(X₁) ⌉`, so
+applying the second rule when the first one does
+not work is equivalent to computing
+```
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ∃ X₁ ⌈ C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+```
+
+Let's see if the two are equivalent:
+
+```
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ∃ X₁ . ⌈ C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// ⌈⌉ commutes with ∃
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ⌈ ∃ X₁ . C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// ∃ X₁ . A ∧ B(X)  ===  A ∧ ∃ X₁ . B(X)
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// ⌈ A ⌉ ∧ P === ⌈ A ∧ P ⌉
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉
+    ⌉
+    ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ (C(Y) ∧ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉)
+    ⌉
+    ∧ ψ₂(X)
+===
+// if A is functional, then A ∧ B = A ∧ ⌈ A ∧ B ⌉
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ (C(Y) ∧ ∃ X₁ . φ₁(X₁))
+    ⌉
+    ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ∧ ¬ ∃ X₁ . φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+```
+
+We can generalize this in the following way:
+
+For a rule `φ(X) ⇒ ψ(X) requires P(X) [priority p]` we take the all
+rules at previous priorities:
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁(X)
+...
+φn(X) ⇒ ψn(X) requires Pn(X)
+```
+and we encode the rule as:
+```
+    φ(X) ∧ P(X)
+    ∧ (¬ ∃ X₁ . φ₁(X₁) ∧ P₁(X₁))
+    ...
+    ∧ (¬ ∃ Xn . φn(Xn) ∧ Pn(Xn))
+    ⇒ psi(X)
+```
+Strictly speaking, we should have the left hand side of the encoding of the
+respective rules under `∃ Xᵢ` above, but the two forms should be
+equivalent.
+
+### Priority predicate
+
+If we have only two rules
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁ [priority(p₁)]
+φ₂(X) ⇒ ψ₂(X) requires P₂ [priority(p₂)]
+```
+and, if we also know that `p₁ < p₂`, we could replace them with
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁
+φ₂(X) ⇒ ψ₂(X)
+    requires P₂(X)
+        ∧ ¬ ∃ X₁ . ⌈φ₂(X) ∧ φ₁(X₁)⌉
```

Comment:
shouldn't `P1` be inside the negation, as well?

Reply:
Fixed.